lib/IDE/Refactoring.cpp

```diff
@@ -2602,6 +2602,96 @@ bool RefactoringActionLocalizeString::performChange() {
   return false;
 }
 
+void generateMemberwiseInit(ClassDecl *classDecl,
+                            SourceEditConsumer &EditConsumer,
+                            SourceManager &SM) {
+  
+  swift::SourceLoc bracesStart = classDecl->getBraces().Start;
+  
+  if (!bracesStart.isValid()) return;
+  
+  swift::SourceLoc targetLocation = bracesStart.getAdvancedLoc(1);
+  
+  if (!targetLocation.isValid()) return;
+  
+  SmallVector<std::string, 5> memberNameVector;
+  SmallVector<std::string, 5> memberTypeVector;
+  
+  for (auto varDecl : classDecl->getStoredProperties()) {
+    auto parentPatternBinding = varDecl->getParentPatternBinding();
+    if (!parentPatternBinding) continue;
+    
+    auto varDeclIndex =
+      parentPatternBinding->getPatternEntryIndexForVarDecl(varDecl);
+    
+    if (auto init = varDecl->getParentPatternBinding()->getInit(varDeclIndex)) {
+      if (init->getStartLoc().isValid())
+        continue;
+    }
+
+    StringRef memberName = varDecl->getName().str();
+    memberNameVector.push_back(memberName.str());
+    
+    std::string memberType = varDecl->getType().getString();
+    memberTypeVector.push_back(memberType);
+  }
+  
+  if (memberTypeVector.size() > 0) {
+    EditConsumer.accept(SM, targetLocation, "\ninternal init(");
+    
+    for (size_t i = 0; i < memberTypeVector.size(); i++) {
+      EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " +
+                          memberTypeVector[i]);
+      
+      if (i != memberTypeVector.size() - 1) {
+        EditConsumer.accept(SM, targetLocation, ", ");
+      }
+    }
+    
+    EditConsumer.accept(SM, targetLocation, ") {\n\n");
+    
+    for (auto varName: memberNameVector) {
+      EditConsumer.accept(SM, targetLocation,
+                          "self." + varName + " = " + varName + "\n");
+    }
+    
+    EditConsumer.accept(SM, targetLocation, "\n}\n");
+  }
+}
+  
+ClassDecl* canGenerateMemberwiseInit(ResolvedCursorInfo CursorInfo) {
+  
+  ClassDecl *classDecl = dyn_cast<ClassDecl>(CursorInfo.ValueD);
+  if (!classDecl || classDecl->getStoredProperties().empty() ||
+      CursorInfo.IsRef) {
+    return nullptr;
+  }
+  
+  return classDecl;
+}
+    
+bool RefactoringActionMemberwiseInitLocalRefactoring::
+  isApplicable(ResolvedCursorInfo Tok, DiagnosticEngine &Diag) {
+  if (Tok.Kind != CursorInfoKind::ValueRef && !Tok.ValueD)
+    return false;
```

Comment:
We can move this if statement to `canGenerateMemberwiseInit` too. `isApplicable` should be a simple wrapper of `canGenerateMemberwiseInit`.